junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/ValidatingVisitor.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2015-2025 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.discovery;
+
+import java.util.function.UnaryOperator;
+
+import org.junit.jupiter.engine.descriptor.Validatable;
+import org.junit.platform.engine.EngineDiscoveryIssue;
+import org.junit.platform.engine.EngineDiscoveryIssue.Severity;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.support.discovery.EngineDiscoveryIssueReporter;
+
+class ValidatingVisitor implements TestDescriptor.Visitor {
+
+	private final EngineDiscoveryIssueReporter issueReporter;
+
+	ValidatingVisitor(EngineDiscoveryIssueReporter issueReporter) {
+		this.issueReporter = issueReporter;
+	}
+
+	@Override
+	public void visit(TestDescriptor descriptor) {
+		if (descriptor instanceof Validatable) {
+			((Validatable) descriptor).validate(new Reporter(descriptor));
+		}
```

Comment:
:speech_balloon: Engines such as Jupiter can use any custom logic to validate during discovery. Here a visitor is used that checks for the Jupiter-specific `Validatable` marker interface.